### PR TITLE
[community] Adding MMR search to YellowBrick vectorDB

### PR DESCRIPTION
**Pull request by Gourang97**
Thank you for contributing to LangChain!

- [ ] **community: add max marginal relevance search to Yellowbrick**: "package: description"
  - **Description:** This PR adds the `max_marginal_relevance_search_with_score_by_vector` method to the `Yellowbrick` class in the community package. This method implements Maximal Marginal Relevance (MMR) for diversity in similarity searches, allowing for more diverse results.
  - **Dependencies:** No additional dependencies required.

- [ ] **Add tests and docs**: If you're adding a new integration, please include:
  1. A test for the `max_marginal_relevance_search_with_score_by_vector` method has been added to the integration test file `test_yellowbrick.py`.

- [ ] **Lint and test**: Integration tests have been added. Formatting and linting done.

If no one reviews your PR within a few days, please @-mention one of baskaryan, efriis, eyurtsev, ccurme, vbarda, hwchase17.

### CODE REVIEW

**Comment by Gourang97** [comment]
@baskaryan, @efriis, @eyurtsev, @ccurme, @vbarda, @hwchase17, could you please take a look and approve

**Comment by Gourang97** [comment]
@baskaryan, this PR has been pending approval since more than 3 weeks, can you please take a look.

File: libs/community/langchain_community/vectorstores/yellowbrick.py [hunk 1]
```diff
@@ -971,3 +971,104 @@ def migrate_schema_v1_to_v2(self) -> None:
...
+        self, 
+        query: str, 
+        k: int = 4, 
+        fetch_k: int = 20, 
+        lambda_mult: float = 0.5,
+        **kwargs: Any
+    ) -> List[Document]:
+        """Return docs selected using the maximal marginal relevance."""
+        embedding = self._embedding.embed_query(query)
+        results = self.max_marginal_relevance_search_with_score_by_vector(
```
Comment (ccurme) [inline libs/community/langchain_community/vectorstores/yellowbrick.py:1038]: (nit) could call `max_marginal_relevance_search_by_vector` here

File: libs/community/langchain_community/vectorstores/yellowbrick.py [hunk 2]
```diff
@@ -971,3 +971,104 @@ def migrate_schema_v1_to_v2(self) -> None:
...
+            List[Document, float]: List of Documents and scores
+        """
+        documents_with_scores = self.similarity_search_with_score_by_vector(
+            embedding=embedding, k=fetch_k, **kwargs
+        )
+        selected_docs = []
+        selected_scores = []
+        while len(selected_docs) < k and documents_with_scores:
+            if not selected_docs:
+                doc, score = documents_with_scores.pop(0)
```
Comment (ccurme) [inline libs/community/langchain_community/vectorstores/yellowbrick.py:1003]: do we know that these are sorted by similarity?

File: libs/community/langchain_community/vectorstores/yellowbrick.py [hunk 3]
```diff
@@ -971,3 +971,104 @@ def migrate_schema_v1_to_v2(self) -> None:
...
+                selected_docs.append(doc)
+                selected_scores.append(score)
+                continue
+            max_score = float("-inf")
+            max_doc = None
+            for doc, score in documents_with_scores:
+                min_similarity = float("inf")
+                for selected_doc in selected_docs:
+                    similarity = self._embedding.similarity(
```
Comment (ccurme) [inline libs/community/langchain_community/vectorstores/yellowbrick.py:1012]: is `.similarity` implemented on all `Embeddings` classes?

File: libs/community/langchain_community/vectorstores/yellowbrick.py [hunk 4]
```diff
@@ -971,3 +971,104 @@ def migrate_schema_v1_to_v2(self) -> None:
                 cursor.execute(insert_content_query)
         except Exception as e:
             raise RuntimeError(f"Failed to migrate schema: {e}") from e
+            
+    def max_marginal_relevance_search_with_score_by_vector(
```
Comment (ccurme) [inline libs/community/langchain_community/vectorstores/yellowbrick.py:975]: (nit) should we implement async methods as well?

File: libs/community/langchain_community/vectorstores/yellowbrick.py [hunk 5]
```diff
@@ -971,3 +971,104 @@ def migrate_schema_v1_to_v2(self) -> None:
...
+            max_score = float("-inf")
+            max_doc = None
+            for doc, score in documents_with_scores:
+                min_similarity = float("inf")
+                for selected_doc in selected_docs:
+                    similarity = self._embedding.similarity(
+                        selected_doc.page_content, doc.page_content
+                    )
+                    min_similarity = min(min_similarity, similarity)
+                mmr_score = lambda_mult * score - (1 - lambda_mult) * min_similarity
```
Comment (ccurme) [inline libs/community/langchain_community/vectorstores/yellowbrick.py:1016]: double-checking that this should be min and not max

**Review by ccurme — COMMENTED** [review]
Left some comments. Feel free to ignore nitpicks.

I'm wondering if we want to use this function (or if requiring numpy is the motivation for skipping it):
https://github.com/langchain-ai/langchain/blob/44e3e2391c48bfd0a8e6a20adde0b6567f4f43c3/libs/core/langchain_core/vectorstores/utils.py#L74

**Pull request closed without merging**